### ISSUE TITLE
false positives SC2154, SC2034 and SC2209 for associative array

### ISSUE THREAD

**Issue by coiby** [koalaman/shellcheck#2399]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154, SC2034 and SC2209 
- My shellcheck version (`shellcheck --version` or "online"): 
```
shellcheck --version          
ShellCheck - shell script analysis tool
version: 0.8.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related
   - #1074 is related but already fixed

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
test_arr()
{
    local -n farr=$1

    farr[c]=cc
    farr[d]=dd
    farr[keb]=dd
    echo "${farr[*]}"
}

declare -A arr

arr[a]=aa
arr[b]=bb
test_arr arr

```

#### Here's what shellcheck currently says:
```

In associate_array.sh line 6:
    farr[c]=cc
         ^-- SC2154 (warning): c is referenced but not assigned.


In associate_array.sh line 7:
    farr[d]=dd
    ^-----^ SC2209 (warning): Use var=$(command) to assign output (or quote to assign string).
         ^-- SC2154 (warning): d is referenced but not assigned.


In associate_array.sh line 8:
    farr[keb]=dd
    ^-------^ SC2209 (warning): Use var=$(command) to assign output (or quote to assign string).
         ^-^ SC2154 (warning): keb is referenced but not assigned.


In associate_array.sh line 16:
arr[ab]=aabb
^-----^ SC2034 (warning): arr appears unused. Verify use (or export if used externally).

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- arr appears unused. Verify use (o...
  https://www.shellcheck.net/wiki/SC2154 -- c is referenced but not assigned.
  https://www.shellcheck.net/wiki/SC2209 -- Use var=$(command) to assign outp...
```

#### Here's what I wanted or expected to see:

empty result

**Comment by petersenna** [comment]
Here is what I have:

```
    VM_EXTRA_OPTS=('--boot-disk-size=40G')

    VM_SELFLINK=$(gcp_vm create --image "$VMTEMPLATE_GCP_IMAGE_URL" \
                --delete-after "$VM_DELETE_TIMEOUT_MINUTES" \
                --extra-opts VM_EXTRA_OPTS)
```
And I got an SC2034 warning about ARRAY.

**Comment by matthewpersico** [comment]
More examples:
```bash
hash[foo]=1
hash[bar]=2
hash[biz]=3
echo '#' "$(set | grep -E '^hash=')"
# hash=([biz]="3" [bar]="2" [foo]="1" )

unset 'hash[foo]'
echo '#' "$(set | grep -E '^hash=')"
# hash=([biz]="3" [bar]="2" )

declare key='bar'
unset 'hash[$key]'
echo '#' "$(set | grep -E '^hash=')"
# hash=([biz]="3" )
```
So this code all worked. But shellecheck.net complained:
```
Line 7:
hash[biz]=3
^-- [SC2034](https://github.com/koalaman/shellcheck/wiki/SC2034) (warning): hash appears unused. Verify use (or export if used externally).
__I think that's wrong; we use it in the unset.__
 
Line 15:
declare key='bar'
        ^-- [SC2034](https://github.com/koalaman/shellcheck/wiki/SC2034) (warning): key appears unused. Verify use (or export if used externally).
__That's definitely wrong; we use it in the unset.__
```

**Comment by netlore** [comment]
I believe I have the same issue:-

`declare -A policy_id
policy_id["dev,rootsvc"]="i-bIe570fT44y"
policy_id["dev,root"]="i-bew870fT42t"
policy_id["prod,rootsvc"]="i-ewe230fT28f"
policy_id["prod,root"]="i-oiu567fT11a"

generates:-

`In manager.sh line 27
policy_id["dev,rootsvc"]="i-bIe570fT44y"
                      ^-- SC2154 dev is referenced but not assigned.

In manager.sh line 29
policy_id["prod,rootsvc"]="i-ewe230fT28f"
                      ^-- SC2154 dev is referenced but not assigned.`

I don't understand why this is reported as an issue, or why only 2 of the 4 lines generate the error...
